doc/compiler/osr/OSR.md

```diff
@@ -20,26 +20,232 @@ OpenJDK Assembly Exception [2].
 SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
 -->
 
+# Introduction
 
-# OSR transition
+On Stack Replacement (OSR) refers to a process of transitioning method execution from compiled to interpreted code while the compiled method is still in progress.
+In particular, it can handle compiled code that has been optimized.
+During this process, the JIT communicates to the VM the correct values of all locals and the operands on the operand stack for
+each interpreter stack frame as if the code had been interpreted before the transition.
+It also informs the VM about the current instruction address.
+The VM then uses all this information to reconstruct all required interpreter stack frames and resume execution
+in the interpreter at the appropriate point.
 
-OSR transition is the process of transfering the execution from jitted code to the interpreter. OSR transition involves leaving the jitted code, reconstructing the runtime VM state and resuming the execution where the jitted code left off in the interpreter.
+# High Level Sequence of Events
 
-# OSR point
-An OSR point is a place in the jitted code where OSR transition is supported.
+OSR is typically triggered by some VM event that applies to a compiled method that is currently on stack.
+For example, such event can be setting a breakpoint in a method on stack.
+In addition, if assumptions that the compiler made while compiling a method are invalidated, OSR might eventually be triggered for the method as well.
 
-At an OSR point, the VM state can be correctly reconstructed and is consistent with the VM state needed by the interpreter to resume the execution at a given bytecode index.
 
-To achieve that, bookkeeping is required at all possible OSR points during ILGen. Due to the cost and complexity of bookkeeping, OSR transition is limited to some well-defined points, which are calls, asyncchecks and monitor enters, etc.
+A VM event can be detected at certain points during execution of a compiled code. These points are called **OSR yield points**. The JIT and the VM agree on the set of the OSR yield points.
 
-Being part of the program, OSR points like calls, asyncchecks and monitor enters can be removed or optimized and no longer be identified as OSR points. However, OSR transition may still be supported at that position.
+Usually, we would like the VM event to be discovered relatively soon after it took place. [Garbage Collection (GC) point](https://github.com/eclipse/omr/blob/81b79405da6c7c960e611a8b2b12fd5861543330/compiler/il/OMRNode.cpp#L2298) meet such criteria.
```

Comment:
I think this should still say *points* (plural)

Reply:
Done